### ISSUE TITLE
JSON-LD Context generator needs container specification

### ISSUE THREAD

**Issue by hsolbrig** [linkml/linkml#181]
The JSON-LD context generator needs to emit `"@container": "@set"` for any multi-valued attribute

**Comment by nlharris** [comment]
Is this still needed? If so, who can it be assigned to?

**Comment by alexbrdn** [comment]
It is, indeed, still needed. I was about to open a new issue on this topic, but I guess it's better to revive this one.

Let me illustrate the problem with an example.

Suppose we have a simple class `Entity` with a multi-valued attribute `hasTag`. When we generate a json-ld context from that, the context contains this definition: 
```json 
    "hasTag": {
      "@id": "local:hasTag"
    }
```

Now, if we use this context to frame an Entity object with two tags, the resulting `hasTag` field is a json array with two values, as you would expect, like this:
```json
 "hasTag": [
    {
      "tag": "t1"
    },
    {
      "tag": "t2"
    }
  ]
```
 [An example in JSON-LD Playground](https://tinyurl.com/48c3d77u) (Note: it appears you have to switch to another tab, e.g. Compacted, and back to Framed when you initially open the page.)

Now, if your Entity only has one tag, when using the same context, the resulting `hasTag` is framed as just that tag object, instead of an array:
```json
  "hasTag": {
    "tag": "t1"
  }
```
See this [example in JSON-LD Playground](https://tinyurl.com/2v29hsyc)

As you can imagine, it is an inconvenience for consumers to deal with a field whose value could be an object or an array of objects.

And the solution to this problem in the JSON-LD world is a simple addition of `"@container": "@set"` to the context, as @hsolbrig suggested. 
If the term is defined in the context as such:
```json
    "hasTag": {
      "@id": "local:hasTag",
      "@container": "@set"
    }
```
then even a single tag value is framed as an array containing that single object:
```json
  "hasTag": [
    {
      "tag": "t1"
    }
  ]
```
Here is [an example](https://tinyurl.com/3r36fh9p) with `@container` in the context. 

Now, coming back around to linkml, it would indeed be very useful if linkml's json-ld context generator would emit `"@container": "@set"` for attributes that have maximum cardinality >1. While it is technically not wrong to omit `@container`, it can be seen as a reasonable default. At the very least, it would be very useful to have an command line option flag to make the linkml generator emit `@container`.

**Comment by Silvanoc** [comment]
@alexbrdn `"@container": "@set"` is only valid for lists/arrays, right? What about dictionaries?

**Comment by Silvanoc** [comment]
@tfliss would you say that the hereby proposed default (use `"@container": "@set"` for lists) is correct? Do you think we need a flag to be able to switch between both approaches? Or simply "hard-coding" the new default suffice?

**Comment by alexbrdn** [comment]
> [@alexbrdn](https://github.com/alexbrdn) `"@container": "@set"` is only valid for lists/arrays, right? What about dictionaries?

Yes. For dictionaries, it should be `"@container": "@index"`, if I am not mistaken. See here: https://www.w3.org/TR/json-ld11/#index-maps
